Angular*CLI.buildArchitectCommand should not fallback to app

The [`AngularBuildCLI.buildArchitectCommand()`](https://github.com/ionic-team/ionic-cli/blob/ionic%405.4.4/packages/ionic/src/lib/project/angular/build.ts#L164) and [`AngularServeCLI.buildArchitectCommand()`](https://github.com/ionic-team/ionic-cli/blob/ionic%405.4.4/packages/ionic/src/lib/project/angular/serve.ts#L268) methods should not hardcode `app` as a fallback project name when no project name is provided via arguments. The best approach would be to read the `defaultProject` key from `angular.json` or by running `ng config defaultProject`

With this, the Ionic CLI will be able to work, by default, in Angular workspaces where the `defaultProject` is not named `app`; as for the ones generated using the `ng add @ionic/angular` schematic; see: #4191

L.E. #4121 should be dealt with first.

You need to use Angular CLI to generate a new workspace with at least an ionic application and some other angular specific library or application. Somewhat related to the fix from here https://github.com/ionic-team/ionic/pull/19765, which was merged but it had a regression that was fixed in https://github.com/ionic-team/ionic/pull/20281.

To reproduce this step you can checkout https://github.com/SchnWalter/ionic-cli-issue-4192 (or follow the "Steps for recreating the demo repository" section below.

If you grep for defaultProject in `angular.json` for a project generated using Ionic CLI, you will see this:
```
> grep defaultProject angular.json
  "defaultProject": "app"
```

But if you do the same after you follow the steps bellow (or just clone the linked repository), you will see this:
```
> grep defaultProject angular.json
  "defaultProject": "my-ionic-client"
```

So, to see the issue you now have to install the latest Ionic CLI (not the modified version)
```
npm install -g @ionic/cli
ionic --version
6.0.1
```

And you will see that the ionic commands will fail:
```
ionic serve
> ng run app:serve --host=localhost --port=8100
[ng] An unhandled exception occurred: Project does not exist.
[ng] See "/tmp/ng-FvRPZs/angular-errors.log" for further details.

[ERROR] ng has unexpectedly closed (exit code 127).

        The Ionic CLI will exit. Please check any output above for error details.
```

Now go back and link the built version that is included in the PR I sent and you will see that the ionic commands have been fixed:
```
ionic serve
> ng run my-ionic-client:serve --host=localhost --port=8100
```


--

## Steps for recreating the demo repository:
```bash
cd ~/MyProjects/

# Create a new empty Angular workspace.
ng new --create-application=false --new-project-root='.' my-frontend-monorepo

cd ~/MyProjects/my-frontend-monorepo/

# Generate two minimal applications, one to be use with ionic.
# We want the ionic application to be the defaultProject, so we generate it first.
ng generate application --minimal --prefix=mic --routing --style=css my-ionic-client
git add . && git commit -m "add my-ionic-client application"
ng generate application --minimal --prefix=mwc --routing --style=css my-web-client
git add . && git commit -m "add my-web-client application"

# Now add Ionic support using the partial Angular CLI schematic
ng add @ionic/angular --project=my-ionic-client
```
And because `ng add @ionic/angular` doesn't create a fully working ionic project (see the first issue above), you need to `ionic.config.json config.xml resources/` from an existing ionic application that was generated using Ionic CLI and also to fix the path to `src/theme/variables.css` which hasn't been fix in the above MRs.
```
cd ~/MyProjects/

# Generate a new temporary Ionic cordova project to extract the missing resources.
ionic start --cordova --type=angular my-ionic-client blank

cd ~/MyProjects/my-ionic-client/

# Copy missing resources to the mono-repository:
# Here I'm using cordova, but I assume that the same would apply to Capacitor.
cp -r config.xml ionic.config.json resources/ src/theme/ ../my-frontend-monorepo

cd ~/MyProjects/my-frontend-monorepo/
git add . && git commit -m "add missing ionic cordova files"

# Delete temporary project
rm -r ~/MyProjects/my-ionic-client/
```

> Ionic CLI properly warns about not knowing which project to use

It only works when you have multiple Ionic applications in the same works. If you have an Angular workspace with multiple projects but only one Ionic application, as the default project, then you get this error:
```
> ionic serve
> ng run app:serve --host=localhost --port=8100
[ng] An unhandled exception occurred: Project does not exist.
[ng] See "/tmp/ng-3sQiH7/angular-errors.log" for further details.

[ERROR] ng has unexpectedly closed (exit code 127).

        The Ionic CLI will exit. Please check any output above for error details.
```
Because there is no Angular project called `app` in `angular.json` and because the `ionic.config.json` doesn't have a projects key with multiple entries.

In multi-app Angular workspaces that contains a single Ionic application, the `ionic.config.json` will most probably look like the one from the [example repository](https://github.com/SchnWalter/ionic-cli-issue-4192/blob/master/ionic.config.json) that I linked in the previous comment:

```
{
  "name": "my-ionic-client",
  "integrations": {
    "cordova": {}
  },
  "type": "angular"
}

```

--- a/packages/@ionic/cli/src/lib/project/angular/__tests__/generate.ts
+++ b/packages/@ionic/cli/src/lib/project/angular/__tests__/generate.ts
@@ -11,7 +11,7 @@ describe('@ionic/cli', () => {
 
         const defaults = {
           name: undefined,
-          project: undefined,
+          project: 'app',
           type: undefined,
         };
 

--- a/packages/@ionic/cli/src/lib/project/angular/__tests__/generate.ts
+++ b/packages/@ionic/cli/src/lib/project/angular/__tests__/generate.ts
@@ -1,0 +1,42 @@
+import { CommandLineOptions } from '../../../../definitions';
+import { AngularGenerateRunner } from '../generate';
+
+describe('@ionic/cli', () => {
+
+  describe('lib/project/angular/generate', () => {
+
+    describe('AngularGenerateRunner', () => {
+
+      describe('createOptionsFromCommandLine', () => {
+
+        const defaults = {
+          name: undefined,
+          project: undefined,
+          type: undefined,
+        };
+
+        it('should provide defaults with no inputs or options', () => {
+          const runner = new AngularGenerateRunner({} as any);
+          const result = runner.createOptionsFromCommandLine([], {} as CommandLineOptions);
+          expect(result).toEqual(defaults);
+        });
+
+        it('should provide options from inputs', () => {
+          const runner = new AngularGenerateRunner({} as any);
+          const result = runner.createOptionsFromCommandLine(['service', 'FancyBar'], { _: [] });
+          expect(result).toEqual({ ...defaults, name: 'FancyBar', type: 'service' });
+        });
+
+        it('should respect --project', () => {
+          const runner = new AngularGenerateRunner({} as any);
+          const result = runner.createOptionsFromCommandLine([], { _: [], project: 'otherProject' });
+          expect(result).toEqual({ ...defaults, project: 'otherProject' });
+        });
+
+      });
+
+    });
+
+  });
+
+});

--- a/packages/@ionic/cli/src/lib/project/angular/__tests__/serve.ts
+++ b/packages/@ionic/cli/src/lib/project/angular/__tests__/serve.ts
@@ -24,7 +24,7 @@ describe('@ionic/cli', () => {
           open: false,
           port: 8100,
           proxy: true,
-          project: undefined,
+          project: 'app',
           prod: undefined,
           platform: undefined,
           verbose: false,
@@ -62,8 +62,8 @@ describe('@ionic/cli', () => {
 
         it('should respect --project and --configuration flags', () => {
           const runner = new AngularServeRunner({} as any);
-          const result = runner.createOptionsFromCommandLine([], { _: [], project: 'app', configuration: 'production' });
-          expect(result).toEqual({ ...defaults, project: 'app', configuration: 'production' });
+          const result = runner.createOptionsFromCommandLine([], { _: [], project: 'otherProject', configuration: 'production' });
+          expect(result).toEqual({ ...defaults, project: 'otherProject', configuration: 'production' });
         });
 
         it('should pass on separated args', () => {

--- a/packages/@ionic/cli/src/lib/project/angular/build.ts
+++ b/packages/@ionic/cli/src/lib/project/angular/build.ts
@@ -80,6 +80,7 @@ ${input('ionic build')} uses the Angular CLI. Use ${input('ng build --help')} to
     const baseOptions = super.createBaseOptionsFromCommandLine(inputs, options);
     const prod = options['prod'] ? Boolean(options['prod']) : undefined;
     const configuration = options['configuration'] ? String(options['configuration']) : (prod ? 'production' : undefined);
+    const project = options['project'] ? String(options['project']) : 'app';
     const sourcemaps = typeof options['source-map'] === 'boolean' ? Boolean(options['source-map']) : undefined;
     const cordovaAssets = typeof options['cordova-assets'] === 'boolean' ? Boolean(options['cordova-assets']) : undefined;
     const watch = typeof options['watch'] === 'boolean' ? Boolean(options['watch']) : undefined;
@@ -87,6 +88,7 @@ ${input('ionic build')} uses the Angular CLI. Use ${input('ng build --help')} to
     return {
       ...baseOptions,
       configuration,
+      project,
       sourcemaps,
       cordovaAssets,
       watch,
@@ -161,8 +163,7 @@ export class AngularBuildCLI extends BuildCLI<AngularBuildOptions> {
 
   protected buildArchitectCommand(options: AngularBuildOptions): string[] {
     const cmd = options.engine === 'cordova' ? 'ionic-cordova-build' : 'build';
-    const project = options.project ? options.project : 'app';
 
-    return ['run', `${project}:${cmd}${options.configuration ? `:${options.configuration}` : ''}`];
+    return ['run', `${options.project}:${cmd}${options.configuration ? `:${options.configuration}` : ''}`];
   }
 }

--- a/packages/@ionic/cli/src/lib/project/angular/generate.ts
+++ b/packages/@ionic/cli/src/lib/project/angular/generate.ts
@@ -100,12 +100,13 @@ To test a generator before file modifications are made, use the ${input('--dry-r
 
   createOptionsFromCommandLine(inputs: CommandLineInputs, options: CommandLineOptions): AngularGenerateOptions {
     const baseOptions = super.createOptionsFromCommandLine(inputs, options);
+    const project = options['project'] ? String(options['project']) : 'app';
 
     // TODO: this is a little gross, is there a better way to pass in all the
     // options that the command got?
     return {
       ...lodash.omit(options, '_', '--', ...GLOBAL_OPTIONS.map(opt => opt.name)),
-      project: options['project'],
+      project,
       ...baseOptions,
     };
   }

--- a/packages/@ionic/cli/src/lib/project/angular/serve.ts
+++ b/packages/@ionic/cli/src/lib/project/angular/serve.ts
@@ -94,6 +94,7 @@ The dev server can use HTTPS via the ${input('--ssl')} option ${chalk.bold.red('
     const prod = options['prod'] ? Boolean(options['prod']) : undefined;
     const ssl = options['ssl'] ? Boolean(options['ssl']) : undefined;
     const configuration = options['configuration'] ? String(options['configuration']) : (prod ? 'production' : undefined);
+    const project = options['project'] ? String(options['project']) : 'app';
     const sourcemaps = typeof options['source-map'] === 'boolean' ? Boolean(options['source-map']) : undefined;
     const consolelogs = typeof options['consolelogs'] === 'boolean' ? Boolean(options['consolelogs']) : undefined;
     const consolelogsPort = consolelogs ? str2num(options['consolelogs-port'], DEFAULT_CONSOLE_LOGS_PORT) : undefined;
@@ -104,6 +105,7 @@ The dev server can use HTTPS via the ${input('--ssl')} option ${chalk.bold.red('
       consolelogsPort,
       ssl,
       configuration,
+      project,
       sourcemaps,
     };
   }
@@ -250,8 +252,7 @@ export class AngularServeCLI extends ServeCLI<AngularServeOptions> {
 
   protected buildArchitectCommand(options: AngularServeOptions): string[] {
     const cmd = options.engine === 'cordova' ? 'ionic-cordova-serve' : 'serve';
-    const project = options.project ? options.project : 'app';
 
-    return ['run', `${project}:${cmd}${options.configuration ? `:${options.configuration}` : ''}`];
+    return ['run', `${options.project}:${cmd}${options.configuration ? `:${options.configuration}` : ''}`];
   }
 }

